### PR TITLE
[mlir] Fix `remove-dead-values` pass throws error when module has a name

### LINKED ISSUE #107870: [mlir] remove-dead-values pass throws error when module has a name
When the input mlir module has a name, like this for example:
```
module @test_opt attributes {} {
  func.func public @test_opt(%arg0: tensor<128x10xf32>) -> tensor<128x10xf32> {
    %0 = tensor.empty() : tensor<64x96xbf16>
    // CHECK-NOT: %[[C:.*]] = tensor.empty[[C:.*]]
    return %arg0 : tensor<128x10xf32>
  }
}
```

then `remove-dead-values` pass throws this error:
```
error: cannot optimize an IR with non-function symbol ops, non-call symbol user ops or branch ops
```

After removing module name (`@test_opt`) everything works. Seems to be a bug.

### PR DESCRIPTION
**Pull request by Wheest**
This fixes the error described in #107870, which I also independently experienced.

This implements the logic change suggested by @joker-eph.

This requires replacing one of the tests:

```mlir
// The IR remains untouched because of the presence of a non-function-like
// symbol op (module @dont_touch_unacceptable_ir).
//
// expected-error @+1 {{cannot optimize an IR with non-function symbol ops, non-call symbol user ops or branch ops}}
module @dont_touch_unacceptable_ir {
  func.func @has_cleanable_simple_op(%arg0 : i32) {
    %non_live = arith.addi %arg0, %arg0 : i32
    return
  }
}
```

Since this asserts that named modules are not allowed, which is the behaviour we are trying to fix.  It is not clear to me why this assertion would be needed.

Anecdotally, I experienced the error going from StableHLO (which has named modules), and lowering.
Fixes #107870.

### CODE REVIEW

File: mlir/test/Transforms/remove-dead-values.mlir [hunk 1]
```diff
@@ -1,13 +1,14 @@
 // RUN: mlir-opt %s -remove-dead-values -split-input-file -verify-diagnostics | FileCheck %s
 
-// The IR remains untouched because of the presence of a non-function-like
-// symbol op (module @dont_touch_unacceptable_ir).
```
Comment (joker-eph) [inline mlir/test/Transforms/remove-dead-values.mlir:4]: Do we have still a test with a non-function symbol op **inside the module**?

**Review by joker-eph — APPROVED** [review]
Looks good, with one thing on the test coverage.

File: mlir/test/Transforms/remove-dead-values.mlir (hunk 1 above)
Comment (Wheest) [reply mlir/test/Transforms/remove-dead-values.mlir:4]: The [2nd test](https://github.com/llvm/llvm-project/blob/eeba53c4d7b815268882ce8d1ed65e9736ef10b9/mlir/test/Transforms/remove-dead-values.mlir#L17-L22) checks for `isa<BranchOpInterface>(op)` and raises the error, but no we don't have a test for the "non-function symbol op inside the module" condition (I don't think we had one for this case before).

In terms of non-function symbol ops, would a global constant make sense? I've added this test which meets the requirements:

```mlir
module {
// expected-error @+1 {{cannot optimize an IR with non-function symbol ops, non-call symbol user ops or branch ops}}
  memref.global "private" constant @__dont_touch_unacceptable_ir : memref<i32> = dense<0>
  func.func @main(%arg0: i32) -> i32 {
    return %arg0 : i32
  }
}
```

**Review by joker-eph — APPROVED** [review]
LG, thanks

**Review by CoTinker — APPROVED** [review]
LGTM.

File: mlir/test/Transforms/remove-dead-values.mlir [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+}
+
+// -----
+
+// Dead values are removed from the IR even if the module has a name
+//
+module @named_module_acceptable {
+  func.func @main(%arg0: tensor<10xf32>) -> tensor<10xf32> {
+    %0 = tensor.empty() : tensor<10xbf16>
+    // CHECK-NOT: %[[C:.*]] = tensor.empty[[C:.*]]
```
Comment (CoTinker) [inline mlir/test/Transforms/remove-dead-values.mlir:21]: Maybe change it to `// CHECK-NOT: tensor.empty` is better.

**Pull request merged**